bcipy/simulator/data/data_engine.py

```diff
@@ -0,0 +1,201 @@
+"""Classes and functions related to loading and querying data to be used in a simulation."""
+import logging
+from abc import ABC, abstractmethod
+from pathlib import Path
+from typing import Any, List, NamedTuple, Union
+
+import numpy as np
+import pandas as pd
+
+from bcipy.helpers.exceptions import TaskConfigurationException
+from bcipy.helpers.parameters import Parameters
+from bcipy.simulator.data import data_process
+from bcipy.simulator.data.data_process import (ExtractedExperimentData,
+                                               RawDataProcessor)
+from bcipy.simulator.helpers.artifact import TOP_LEVEL_LOGGER_NAME
+
+log = logging.getLogger(TOP_LEVEL_LOGGER_NAME)
+
+
+class Trial(NamedTuple):
+    """Data for a given trial (a symbol within an Inquiry).
+
+    Attrs
+    -----
+        source - directory of the data source
+        inquiry_n - starts at 0; does not reset at each series
+        inquiry_pos  - starts at 1; position in which the symbol was presented
+        symbol - alphabet symbol that was presented
+        target - 1 or 0 indicating a boolean of whether this was a target symbol
+        eeg - EEG data associated with this trial
+    """
+    source: str
+    # series_n: int
+    inquiry_n: int
+    inquiry_pos: int
+    symbol: str
+    target: int
+    eeg: np.ndarray  # Channels by Samples ; ndarray.shape = (channel_n, sample_n)
+
+    def __str__(self):
+        fields = [
+            f"source='{self.source}'", f"inquiry_n={self.inquiry_n}",
+            f"inquiry_pos={self.inquiry_pos}", f"symbol='{self.symbol}'",
+            f"target={self.target}", f"eeg={self.eeg.shape}"
+        ]
+        return f"Trial({', '.join(fields)})"
+
+    def __repr__(self):
+        return str(self)
+
+
+class QueryFilter(NamedTuple):
+    """Provides an API used to query a data engine for data."""
+    field: str
+    operator: str
+    value: Any
+
+    def is_valid(self) -> bool:
+        """Check if the filter is valid."""
+        # pylint: disable=no-member
+        return self.field in Trial._fields and self.operator in self.valid_operators and isinstance(
+            self.value, Trial.__annotations__[self.field])
+
+    @property
+    def valid_operators(self) -> List[str]:
+        """List of supported query operators"""
+        return ["<", "<=", ">", ">=", "==", "!="]
+
+
+class DataEngine(ABC):
+    """Abstract class for an object that loads data from one or more sources,
+    processes the data using a provided processor, and provides an interface
+    for querying the processed data."""
+
+    def load(self):
+        """Load data from sources."""
+
+    @property
+    def trials_df(self) -> pd.DataFrame:
+        """Returns a dataframe of Trial data."""
+
+    @abstractmethod
+    def query(self,
+              filters: List[QueryFilter],
+              samples: int = 1) -> List[Trial]:
+        """Query the data."""
+
+
+def convert_trials(data_source: ExtractedExperimentData) -> List[Trial]:
```

Comment:
Can this method also provide a list of Inquiries? Or, can it leave the structure as-is if the data is extracted by inquiries already? (Inquiries would later be useful for gaze simulation)